addons/mail/static/src/messaging/entity/attachment/attachment.js

```diff
@@ -0,0 +1,305 @@
+odoo.define('mail.messaging.entity.Attachment', function (require) {
+'use strict';
+
+const { registerNewEntity } = require('mail.messaging.entity.core');
+
+function AttachmentFactory({ Entity }) {
+
+    let nextTemporaryId = -1;
+    function getAttachmentNextTemporaryId() {
+        const id = nextTemporaryId;
+        nextTemporaryId -= 1;
+        return id;
+    }
+    class Attachment extends Entity {
+
+        //----------------------------------------------------------------------
+        // Public
+        //----------------------------------------------------------------------
+
+        /**
+         * @param {string} filename
+         * @returns {mail.messaging.entity.Attachment[]}
+         */
+        static AllFromFilename(filename) {
+            const allAttachments = this.all;
+            return allAttachments.filter(attachment => attachment.filename === filename);
+        }
+
+        /**
+         * @param {string} filename
+         * @returns {mail.messaging.entity.Attachment|undefined}
+         */
+        static temporaryFromFilename(filename) {
+            const attachments = this.AllFromFilename(filename);
+            return attachments.find(attachment => attachment.isTemporary);
+        }
+
+        /**
+         * View provided attachment(s), with given attachment initially. Prompts
+         * the attachment viewer.
+         *
+         * @param {Object} param0
+         * @param {mail.messaging.entity.Attachment} [param0.attachment]
+         * @param {mail.messaging.entity.Attachment>[]} param0.attachments
+         * @returns {string|undefined} unique id of open dialog, if open
+         */
+        static view({ attachment, attachments }) {
+            if (!attachments) {
+                return;
+            }
+            if (!attachment) {
+                attachment = attachments[0];
```

Comment:
I would test attachments.length above, I don't like to implicitly rely on [0] being undefined if not found.